test/FunctionalTests/Microsoft.OData.Core.Tests/Json/MockStreamBasedJsonWriterFactoryWrapper.cs

```diff
@@ -0,0 +1,57 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="MockStreamBasedJsonWriterFactoryWrapper.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP3_1_OR_GREATER
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.OData.Json;
+
+namespace Microsoft.OData.Tests.Json
+{
+    /// <summary>
+    /// This wraps an <see cref="IStreamBasedJsonWriterFactory"/> to
+    /// intercept calls made to create a writer in order to allow
+    /// testing code paths around JSON writer construction.
+    /// </summary>
+    internal class MockStreamBasedJsonWriterFactoryWrapper : IStreamBasedJsonWriterFactory

```

Comment:
```suggestion
    internal sealed class MockStreamBasedJsonWriterFactoryWrapper : IStreamBasedJsonWriterFactory
```

Reply:
Is it beneficial to seal test classes?

Reply:
I think it's always beneficial to seal classes. Test classes in particular are really easy to *unsealed* so I don't see a reason not to.

Reply:
> 

I know I'm a rando and don't have skin in the game on this particular issue, but it's almost never "beneficial" to seal classes. You NEVER know what people are going to want to do with a class, and the ways to access functionality in a sealed class are almost always _more_ detrimental than just keeping the class public.

That is doubly true when it comes to OData. I've run into a TON of issues building Restier that are made more difficult by sealed classes and functionality that is not replaceable.

Just something to think about. Good luck with your testing!

Reply:
I have sealed the classes. But out of curiosity, is this a common or standard practice? I haven't seen that in other framework code (including AspNetCore and BCL code).

Reply:
@robertmclaws I think that's a good point. For maintainability it's usually easier to go from `sealed` to `unsealed` (non-breaking change) than the other way around. Also when something is sealed, it makes it easier to make assumptions about how it's being used and opens up room for more optimizations. But it comes at a great cost for extensibility so maybe it's something when the benefits outweigh the cost.

Reply:
> I have sealed the classes. But out of curiosity, is this a common or standard practice? I haven't seen that in other framework code (including AspNetCore and BCL code).

:) I feel like your response to Robert answers your own question :D

@robertmclaws sorry, just now following up on this. It's definitely something to think about. What I've noticed is that most classes that aren't `sealed` don't have any virtual members anyway. So, deriving from them doesn't really provide much value, and prevents us from making assumptions about them (as Clement mentions). There are times when it's useful to derive a class without virtual members, usually in cases where type-safety is not guaranteed (dependency injection, serialization, tree traversals, etc.). I tend to default to a `sealed` class, but I definitely understand your frustration since OData makes pretty extensive use of those other cases. I'm still trying to really define the line of when it's a good idea or not.